src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,761 @@
+		int rc2 = dfs_release(obj);
+		if (rc2) {
+			D_ERROR("Could not release object '%s': %s (rc=%d)", path, strerror(rc2),
+				rc2);
+		}
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+/* describes the IO operation */
+struct io_op {
+	daos_event_t ev;
+
+	dfs_obj_t   *obj;
+	PyObject    *item;
+
+	/* Purely for debug purpose: should not be freed as it's not the owner of the data */
+	const char  *path;
+
+	daos_size_t  size;
+	int          err;
+
+	d_iov_t      iov;
+	d_sg_list_t  sgl;
+
+	/* Buffer view that implements Python Buffer Protocol, has to be release after use */
+	Py_buffer    buf_view;
+};
+
+static int
+start_read_op(struct dfs_handle *hdl, PyObject *item, struct io_op *op)
+{
+	assert(op != NULL);
+
+	int           rc  = 0;
+	daos_event_t *evp = &op->ev;
+
+	PyObject     *py_path = PyTuple_GetItem(item, 0);
+	PyObject     *py_buff = PyTuple_GetItem(item, 1);
+
+	if (py_path == NULL || py_buff == NULL) {
+		D_WARN("Each tuple must contain exactly two elements: path and bytearray");
+		return EINVAL;
+	}
+
+	const char *path = PyUnicode_AsUTF8(py_path);
+	if (path == NULL) {
+		D_WARN("First element of a tuple does not look like a path");
+		return EINVAL;
+	}
+
+	if (PyObject_GetBuffer(py_buff, &op->buf_view, PyBUF_WRITE) == -1) {
+		D_WARN("Buffer is not writable");
+		return EINVAL;
+	}
+
+	if (!PyBuffer_IsContiguous(&op->buf_view, 'C')) {
+		D_WARN("Buffer for '%s' is not contiguous", path);
+		rc = EINVAL;
+		goto out;
+	}
+
+	rc = daos_event_init(evp, hdl->eq, NULL);
+	if (rc) {
+		D_ERROR("Could not init event: %s (rc=%d)", d_errstr(rc), rc);
+		rc = daos_der2errno(rc);
+		goto out;
+	}
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &op->obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup path '%s': %s (rc=%d)", op->path, strerror(rc), rc);
+		goto out;
+	}
+
+	op->path = path;
+	op->size = op->buf_view.len;
+	d_iov_set(&op->iov, op->buf_view.buf, op->size);
+
+	op->sgl.sg_nr     = 1;
+	op->sgl.sg_nr_out = 0;
+	op->sgl.sg_iovs   = &op->iov;
+
+	rc = dfs_read(hdl->dfs, op->obj, &op->sgl, 0 /* offset */, &op->size, &op->ev);
+	if (rc) {
+		D_ERROR("Could not start async read on '%s': %s (rc=%d)", op->path, strerror(rc),
+			rc);
+		goto out;
+	}
+
+	return 0;
+
+out:
+	PyBuffer_Release(&op->buf_view);
+
+	int rc2 = daos_event_fini(&op->ev);
+	if (rc2) {
+		D_ERROR("Could not finalize event: %s (rc=%d)", d_errstr(rc2), rc2);
+	}
+
+	if (op->obj != NULL) {
+		rc2 = dfs_release(op->obj);
+		if (rc2) {
+			D_ERROR("Could not release obj handler of '%s': %s (rc=%d)", op->path,
+				strerror(rc2), rc2);
+		}
+		op->obj = NULL;
+	}
+
+	return rc;
+}
+
+static int
+complete_read_op(struct dfs_handle *hdl, struct io_op *op)
+{
+	assert(op != NULL);
+
+	D_DEBUG(DB_ANY, "READ of %zu bytes from '%s' completed with  status: %s (rc = %d)",
+		op->size, op->path, d_errstr(op->ev.ev_error), op->ev.ev_error);
+
+	int rc = dfs_release(op->obj);
+	if (rc) {
+		D_WARN("Could not release object handler %s: %s (rc=%d)", op->path, strerror(rc),
```

Comment:
these all should not be warnings but actual errors, that you should not ignore the rc of